server/src/main/java/io/druid/segment/loading/OmniSegmentLoader.java

```diff
@@ -127,22 +136,27 @@ public File getSegmentFiles(DataSegment segment) throws SegmentLoadingException
           log.debug("Unable to make parent file[%s]", storageDir);
         }
         try {
-          downloadStartMarker.createNewFile();
+          if(!downloadStartMarker.createNewFile()){
+            // TODO: should this fail?
+            log.warn("Was not able to create new download marker for [%s]", storageDir);
+          }
         }
         catch (IOException e) {
-          throw new SegmentLoadingException("Unable to create marker file for [%s]", storageDir);
+          throw new SegmentLoadingException(e, "Unable to create marker file for [%s]", storageDir);
         }
       }
 
-      getPuller(segment.getLoadSpec()).getSegmentFiles(segment, storageDir);
+      if(segment.getLoadSpec().containsKey("uri")){
+        loadURI(segment.getLoadSpec(), storageDir);
+      } else {
+        // legacy
+        getPuller(segment.getLoadSpec()).getSegmentFiles(segment, storageDir);
+      }
 
       if (!downloadStartMarker.delete()) {
         throw new SegmentLoadingException("Unable to remove marker file for [%s]", storageDir);
       }
 
-
-      loc.addSegment(segment);
```

Comment:
Looked deeper, I think it is a bit redundant but doesn't double-account for segments.